src/gc/gc.d

```diff
@@ -1543,6 +1519,13 @@ struct Gcx
      */
     void runFinalizers(in void[] segment) nothrow
     {
+        GC.inFinalizer = true;
+        scope (failure) // unlock on finalize errors
+        {
+            GC.inFinalizer = false;
+            GC.gcLock.unlock();
```

Comment:
I changed everything to use `runLocked` and added a `scope (failure)` there.
This will cost a few cycles, b/c dmd can't currently inline a function using exception handling.